pkg/common/controlplaneendpoint/virtualip/fromreader.go

```diff
@@ -0,0 +1,48 @@
+// Copyright 2023 D2iQ, Inc. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package virtualip
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/api/v1alpha1"
+)
+
+type fromReaderProvider struct {
+	template string
+}
+
+func NewFromReaderProvider(reader io.Reader) (*fromReaderProvider, error) {
+	buf := new(strings.Builder)
+	_, err := io.Copy(buf, reader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to copy content from reader: %w", err)
+	}
+
+	return &fromReaderProvider{
+		template: buf.String(),
+	}, nil
+}
+
+func (p *fromReaderProvider) GetFile(
+	_ context.Context,
+	spec v1alpha1.ControlPlaneEndpointSpec,
```

Comment:
should we make this a kube-vip config structure with one of the value as CPEndpoint. we can set default values if needed but all the parameters are technically modifiable. sure capx does not support it yet but here we can allow to modify any parameters of kube-vip

Reply:
@deepakm-ntnx sorry I'm not sure if understand the question, can you please provide and example.

Reply:
what i meant was technically all of the following env var which we set in the yaml are variables and user can change them based on need. we can make it all variables in capx as well (we already have a tracker). we can make it variables in here as well (could be with default values to start with) so that say user can update svc_enable etc
<pre>
            - name: vip_arp
              value: "true"
            - name: address
              value: "{{ `{{ .ControlPlaneEndpoint.Host }}` }}"
            - name: port
              value: "{{ `{{ .ControlPlaneEndpoint.Port }}` }}"
            - name: vip_cidr
              value: "32"
            - name: cp_enable
              value: "true"
            - name: cp_namespace
              value: kube-system
            - name: vip_ddns
              value: "false"
            - name: vip_leaderelection
              value: "true"
            - name: vip_leaseduration
              value: "15"
            - name: vip_renewdeadline
              value: "10"
            - name: vip_retryperiod
              value: "2"
            - name: svc_enable
              value: "false"
            - name: lb_enable
              value: "false"
            - name: enableServicesElection
              value: "false"
</pre>

Reply:
https://jira.nutanix.com/browse/KRBN-8114

Reply:
Ahh got it. With https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/558/commits/c8c9c176801bdafafe13188382525fd5ae63d062 I've extended an API where in the future we can support additional configuration. IMO we should wait and only expose the configuration that we are ready to test/support.

For now though, its possible to override the full Pod spec by changing the ConfigMap. (similar pattern to the addons)

Reply:
Thinking about this, I wonder if this is better to add kube-vip as an addon API rather than under controlplane here? This would allow for configuring kube-vip service LB, etc rather than just saying it's for CP.

Reply:
kubevip pod has to start along with first control plane for capi controllers to connect with api server. 
configuring kube-vip as service LB is a good idea, but i am not sure if we should use the same kube-vip pods for CP.